lib/Support/JSON.cpp

```diff
@@ -0,0 +1,141 @@
+//===- Json.cpp - Json Utilities --------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utilities for file system path handling, supplementing the ones from
+// llvm::sys::path.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/JSON.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace json = llvm::json;
+
+using namespace circt;
+using mlir::UnitAttr;
+
+static bool fitsInDouble(const APInt &apint) {
+  if (apint.getActiveBits() >= std::numeric_limits<int64_t>::digits)
+    return false;
+  int64_t value = apint.getSExtValue();
+  return value == static_cast<int64_t>(static_cast<double>(value));
+}
+
+// NOLINTBEGIN(misc-no-recursion)
+LogicalResult circt::convertAttributeToJSON(llvm::json::OStream &json,
+                                            Attribute attr) {
+  return TypeSwitch<Attribute, LogicalResult>(attr)
+      .Case<DictionaryAttr>([&](auto attr) {
+        json.objectBegin();
+        for (auto subAttr : attr) {
+          json.attributeBegin(subAttr.getName());
+          if (failed(convertAttributeToJSON(json, subAttr.getValue())))
+            return failure();
+          json.attributeEnd();
+        }
+        json.objectEnd();
+        return success();
+      })
+      .Case<ArrayAttr>([&](auto attr) {
+        json.arrayBegin();
+        for (auto subAttr : attr)
+          if (failed(convertAttributeToJSON(json, subAttr)))
+            return failure();
+        json.arrayEnd();
+        return success();
+      })
+      .Case<BoolAttr, StringAttr>([&](auto attr) {
+        json.value(attr.getValue());
+        return success();
+      })
+      .Case<IntegerAttr>([&](auto attr) -> LogicalResult {
+        // If the integer can be accurately represented by a double, print
+        // it as an integer. Otherwise, convert it to an exact decimal string.
+        const auto &apint = attr.getValue();
+        if (!fitsInDouble(apint))
+          return failure();
```

Comment:
I'm still unclear on this. Is there a reason why we care if an integer is representable by a double? That seems to be overlaying Javascript internals onto MLIR internals while JSON (at least the ECMA-404 standard) ignores the implementation of the programming language:

> JSON is agnostic about the semantics of numbers. In any programming language, there can be a variety of number types of various capacities and complements, fixed or floating, binary or decimal. That can make interchange between different programming languages difficult. JSON instead offers only the representation of numbers that humans use: a sequence of digits. All programming languages know how to make sense of digit sequences even if they disagree on internal representations. That is enough to allow interchange.

Given that, I think it's fine to remove the `fitsInDouble` check and just store it as a number which may not be representable as an `int64_t` or as a number that precisely lines up with a `double`. If the number isn't representable in a later language that reads this, e.g., Javascript, then it's the responsibility of that language to handle it accordingly using non-double-based libraries.

WDYT?

Reply:
Thanks for spotting this - I thought JSON was so tied to JS it inherited the silly number semantics. I'll remove this stuff.